synapse/handlers/_base.py

```diff
@@ -133,10 +133,9 @@ def handle_new_client_event(self, event, context, extra_destinations=[],
         for k, s in context.current_state.items():
             try:
                 if k[0] == EventTypes.Member:
-                    if s.content["membership"] == Membership.JOIN:
```

Comment:
I was being a muppet. Turns out `persist_event` was modifying `context.current_state` to add the ban event to the current state, removing the previous join. I've fixed this so that the correct destinations are added.

Ftr, `context.current_state` is a dict from the tuple `(type, state_key)` to `FrozenEvent`